### PR TITLE
Add copy() method to ndarray

### LINKED ISSUE #980: Add copy() method to ndarray
**Describe the bug**
The `copy()` function is very useful, but currently does not exist in pyccel. It should be added

**Test Code**
Provide code which does not currently work but which should do when this issue is fixed:
```python
r = s.copy()
```
This should be translated to:
```python
r = np.empty_like(s)
r[:] = s
```

### PR DESCRIPTION
**Pull request by EmilyBourne**
Use the existing implementation of `NumpyArray` to add support for `copy`. In the case of the class method the default arguments differ from the `NumpyArray` arguments. This is achieved by adding the `argument_description` parameter to `PyccelFunctionDef`. Fixes #980

### CODE REVIEW

**Comment by yguclu** [comment]
Who should review this pull request into the branch of PR #1256?

File: docs/numpy-functions.md [hunk 1]
```diff
@@ -564,9 +564,11 @@ In Pyccel we try to support the NumPy functions which developers use the most..
 
     -   `empty`, `full`, `ones`, `zeros`, `array`, `arange` (`like` parameter is not supported).
     -   `empty_like`, `full_like`, `zeros_like`, and `ones_like` (`subok` parameter is not supported).
+    -   `array` (`copy`, `subok`, and `like` parameter is not supported).
```
Comment (yguclu) [inline docs/numpy-functions.md:567]: Typo:
```suggestion
    -   `array` (`copy`, `subok`, and `like` parameters are not supported).
```

File: pyccel/ast/core.py [hunk 2]
```diff
@@ -2924,25 +2924,42 @@ class PyccelFunctionDef(FunctionDef):
...
+
+    @property
+    def argument_description(self):
+        """
+        Get a description of the arguments.
+
+        If the underlying class has different default values, return a dictionary
+        containing all arguments and their default values. Otherwise return `None`.
+        """
+        return self._argument_description
```
Comment (yguclu) [inline pyccel/ast/core.py:2962]: I think that `self._argument_description` can be an empty dict `{}`, but never `None`. See `__init__`.

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 3]
```diff
@@ -5986,3 +5984,56 @@ def expression_type():
...
+        return b
+
+    @template('T', ['float[:,:]', 'complex[:,:,:](order=F)'])
+    def copy_array_to_F(a : 'T'):
+        b = a.copy(order='F')
+        return b
+
+    @template('T', ['float[:,:]', 'complex[:,:,:](order=F)'])
+    def copy_array_to_C(a : 'T'):
+        b = a.copy(order='F')
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:6010]: Typo:
```suggestion
        b = a.copy(order='C')
```

**Review by yguclu — CHANGES_REQUESTED** [review]

File: pyccel/ast/core.py [hunk 4]
```diff
@@ -2924,25 +2924,42 @@ class PyccelFunctionDef(FunctionDef):
...
+
+    argument_description : dict, optional
+        A dictionary containing all arguments and their default values. This
+        is useful in order to reuse types with similar functionalities but
+        different default values.
     """
-    __slots__ = ()
-    def __init__(self, name, func_class, *, decorators = {}):
+    __slots__ = ('_argument_description',)
+    def __init__(self, name, func_class, *, decorators = {}, argument_description = None):
```
Comment (EmilyBourne) [inline pyccel/ast/core.py:2943]:
```suggestion
    def __init__(self, name, func_class, *, decorators = {}, argument_description = {}):
```

File: pyccel/ast/core.py [hunk 5]
```diff
@@ -2924,25 +2924,42 @@ class PyccelFunctionDef(FunctionDef):
...
         assert isinstance(func_class, type) and \
                 issubclass(func_class, (PyccelInternalFunction, PyccelAstNode))
+        assert argument_description is None or isinstance(argument_description, dict)
```
Comment (EmilyBourne) [inline pyccel/ast/core.py:2946]:
```suggestion
        assert isinstance(argument_description, dict)
```

File: pyccel/ast/core.py [hunk 6]
```diff
@@ -2924,25 +2924,42 @@ class PyccelFunctionDef(FunctionDef):
...
         self._cls_name = func_class
+        self._argument_description = argument_description or {}
+
+    @property
+    def argument_description(self):
+        """
+        Get a description of the arguments.
+
+        If the underlying class has different default values, return a dictionary
+        containing all arguments and their default values. Otherwise return `None`.
```
Comment (EmilyBourne) [inline pyccel/ast/core.py:2960]:
```suggestion
        Return a dictionary whose keys are the arguments with default values
        and whose values are the default values for the function described by
        the `PyccelFunctionDef`
```

File: pyccel/parser/semantic.py [hunk 7]
```diff
@@ -892,9 +892,17 @@ def _handle_function(self, expr, func, args, is_method = False):
...
+            argument_description = func.argument_description
             func = func.cls_name
             args, kwargs = split_positional_keyword_arguments(*args)
 
+            if argument_description:
+                # Ignore values passed by position but add any unspecified keywords
+                # with the correct default value
+                for kw, val in list(argument_description.items())[len(args):]:
+                    if kw not in kwargs:
+                        kwargs[kw] = val
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:904]:
```suggestion
            # Ignore values passed by position but add any unspecified keywords
            # with the correct default value
            for kw, val in list(argument_description.items())[len(args):]:
                if kw not in kwargs:
                    kwargs[kw] = val
```

**Review by yguclu — CHANGES_REQUESTED** [review]
Looks good, but maybe there is one small improvement to make.

File: pyccel/ast/core.py [hunk 8]
```diff
@@ -2924,25 +2924,43 @@ class PyccelFunctionDef(FunctionDef):
...
+    def __init__(self, name, func_class, *, decorators = {}, argument_description = {}):
         assert isinstance(func_class, type) and \
                 issubclass(func_class, (PyccelInternalFunction, PyccelAstNode))
+        assert isinstance(argument_description, dict)
         arguments = ()
         results = ()
         body = ()
         super().__init__(name, arguments, results, body, decorators=decorators)
         self._cls_name = func_class
+        self._argument_description = argument_description or {}
```
Comment (EmilyBourne) [inline pyccel/ast/core.py:2952]:
```suggestion
        self._argument_description = argument_description
```

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/core.py]: Opps. Fixed

**Review by yguclu — APPROVED** [review]

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 9]
```diff
@@ -5986,3 +5984,56 @@ def expression_type():
...
+    def copy_array_to_F(a : 'T'):
+        b = a.copy(order='F')
+        return b
+
+    @template('T', ['float[:,:]', 'complex[:,:,:](order=F)'])
+    def copy_array_to_C(a : 'T'):
+        b = a.copy(order='C')
+        return b
+
+    arr_1d = randint(min_int64, max_int64, size=5)
```
Comment (EmilyBourne) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:6013]:
```suggestion
    arr_1d = randint(min_int, max_int, size=5)
```

**Review by yguclu — APPROVED** [review]

**Pull request merged**